Initial import of node web api accessing another protected api as the user (#2)
Azure-Samples/ms-identity-docs-code-javascript: protected-api-access-protected-api/README.md

<!-- Keeping yaml frontmatter commented out for now
---
# Metadata required by https://docs.microsoft.com/samples/browse/
# Metadata properties: https://review.docs.microsoft.com/help/contribute/samples/process/onboarding?branch=main#add-metadata-to-readme
languages:

> - Node.Js

Reviewer: Just make sure all `.Js` -> `.js` as was requested, in the README here.

<!-- Keeping yaml frontmatter commented out for now
---
# Metadata required by https://docs.microsoft.com/samples/browse/
# Metadata properties: https://review.docs.microsoft.com/help/contribute/samples/process/onboarding?branch=main#add-metadata-to-readme
languages:
- Node.js
page_type: sample
name: "Node.Js application that makes a request to the Graph API from a protected API on behalf of a user"
description: "This sample Node.Js application shows a confidential client application which calls a protected API which make a request to Microsoft Graph using the on-behalf-of flow."
products:
- azure
- azure-active-directory
- ms-graph

> urlFragment: ms-identity-docs-code-api-nodejs

Author: @ckittel is the double dash before "obo" intentional or should it just be a single dash?

> urlFragment: ms-identity-docs-code-api-nodejs

Reviewer: single, sorry, typo on my side.

> urlFragment: ms-identity-docs-code-api-nodejs

Author: this one should be resolved

---
-->

# Node.js | web API | user sign-in, access control (protected routes), protected web API access (Microsoft Graph) | Microsoft identity platform

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This sample Node.js application demonstrates a confidential client application which calls a protected API which then makes a request to Microsoft Graph via the on-behalf-of flow.

> 

Reviewer: ~~~suggestion

```console
$ curl http://localhost:8080/me -H "Authorization: Bearer {valid-access-token}"
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users/$entity",
  "businessPhones": ["+1 (999) 5551001"],
  "displayName": "Contoso Employee",
  "givenName": "Contoso",
  "jobTitle": "Worker",
  "mail": "cemployee@contoso.com",
  "mobilePhone": "1 999-555-1001",
  "officeLocation": "Contoso Plaza/F30",
  "preferredLanguage": null,
  "surname": "Employee",
  "userPrincipalName": "contoso_employee@contoso.com",
  "id": "e3a49d8b-d849-48eb-9947-37c1f9589812"
}

```
~~~

|:---------------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
| **Name**                          | `Node Web API`                                                               | Suggested value for this sample. <br/> You can change the app name at any time.                    |
| **Supported account types**       | **Accounts in this organizational directory only (Single tenant)**           | Suggested value for this sample.                                                                   |
| **Platform type**                 | _None_                                                                      | No redirect URI required; don't select a platform.                                                                    |
| **Client secret**                 | _**Value** of the client secret (not its ID)_                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |

> | **API Permissions**               | `user_impersonation`                                                         | Create a new permission called user_impersonation.  Required value for this sample.      |
> 
> 
> > :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

Reviewer: ```suggestion
| **API Permissions**               | `user_impersonation`                                                         | Create a new permission called user_impersonation.  Required value for this sample.      |

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.
```

> 
> 
> 
> ### 2. Update code sample with app registration values

Reviewer: ```suggestion

### 2. Update code sample with app registration values
```

```javascript
  auth: {
    // 'Application (client) ID' of app registration in Azure portal - this value is a GUID
    clientId: '',
    
    // Client secret 'Value' (not the ID) from 'Client secrets' in app registration in Azure portal
    clientSecret: '',
    
    // Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
    authority: ''
  }
```

```javascript
  secret: jwks.expressJwtSecret({
    // Full URL, in the form of: https://login.microsoftonline.com/<tenant>/discovery/v2.0/keys
    jwksUri: ''
  }),
```

> 

Reviewer: Assuming you apply the app change I suggested.

```suggestion

```

### 3. Install package(s)

> To install Node.Js and MSAL libraries:

Reviewer: ```suggestion
To install Node.js and MSAL libraries:
```

```bash
npm install
```

## Run the application

```bash
node app.js
```

## Browse to the application

> Using postman or a similar app, navigate to **http://localhost:8080/me**

Reviewer: Using Postman, curl, or a similar application, issue an HTTP GET request to *http://localhost:8080/me* with an `Authorization` header of `Bearer {VALID-ACCESS-TOKEN}`.

Author: Updated

> You will need to provide an HTTP GET request such as:
> 
> Authorization: Bearer {token}
> 
> 
> If everything worked, the sample app should produce the JSON-formatted output of a graph /me request for your user.

Reviewer: ```suggestion
For example, if you use curl and everything worked, the sample you should receive a response from the API similar to this:
```

```console

> {
>   "@odata.context":"https://graph.microsoft.com/v1.0/$metadata#users/$entity",
>   "businessPhones":[],
>   "displayName":"Firstname Lastname",
>   "givenName":"",
>   "jobTitle":null,
>   "mail":null,
>   "mobilePhone":null,
>   "officeLocation":null,
>   "preferredLanguage":null,
>   "surname":"Lastname",
>   "userPrincipalName":"",
>   "id":"00000000-0000-0000-0000-000000000000"
>   }

Reviewer: ```suggestion
$ curl http://localhost:8080/me -H "Authorization: Bearer {VALID-ACCESS-TOKEN}"
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users/$entity",
  "businessPhones": ["+1 (999) 5551001"],
  "displayName": "Contoso Employee",
  "givenName": "Contoso",
  "jobTitle": "Worker",
  "mail": "cemployee@contoso.com",
  "mobilePhone": "1 999-555-1001",
  "officeLocation": "Contoso Plaza/F30",
  "preferredLanguage": null,
  "surname": "Employee",
  "userPrincipalName": "contoso_employee@contoso.com",
  "id": "e3a49d8b-d849-48eb-9947-37c1f9589812"
}
```

```

> 
> 
> 
> ## About the code

Reviewer: ```suggestion

### Generating a valid access token

Follow the instructions in [Gaining consent for the middle-tier application](https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#gaining-consent-for-the-middle-tier-application), including setting the API's app registration manifest value of **knownClientApplications**.

## About the code
```

This demo shows how a confidential client application in Node.Js can call a protected web API on-behalf-of the initiating user.

> 

Reviewer: ```suggestion
This Node.js application uses the Express web framework. The app has a single route that requires an access token. The access token will be automatically validated by MSAL:

- A missing or invalid (expired, wrong audience, etc) token will result in a `401` response.
- An otherwise valid token without the proper scope will result in a `403` response.
- A valid token with the proper scope of `user_impersonation` will be accepted, and the token will then be exchanged in an on-behalf-of OAuth flow to call to Microsoft Graph. The results of the Microsoft Graph call are then returned as the results of the API call.
```

Author: this one should be resolved

- Node.Js
page_type: sample
name: "Node.Js application that makes a request to the Graph API from a protected API on behalf of a user"
description: "This sample Node.Js application shows a confidential client application which calls a protected API which make a request to Microsoft Graph using the on-behalf-of flow."
products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-api-nodejs
---
-->

# Node.js | web app | user sign-in, access control (protected routes), protected web API access (Microsoft Graph) | Microsoft identity platform

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This sample Node.Js application demonstrates a confidential client application which calls a protected API which then makes a request to Microsoft Graph via the on-behalf-of flow.

## Prerequisites

- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
- Node.Js 16+
- An existing Web API

## Setup

> ### 1. Modify your existing Web API

Reviewer 2: Instead of an existing web API, should they not register the web API?

Author: Updating Readme, this part shouldn't have been included.

| **Name**                          | `Node Web APP`                                                               | Suggested value for this sample. <br/> You can change the app name at any time.                    |
| **Supported account types**       | **Accounts in this organizational directory only (Single tenant)**           | Suggested value for this sample.                                                                   |
| **Platform type**                 | **Web**                                                                      | Required value for this sample.                                                                    |
| **Redirect URI**                  | `http://localhost:8080/redirect`                                             | Required value for this sample.                                                                    |
| **Client secret**                 | _**Value** of the client secret (not its ID)_                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |

> | **API Permissions**               | `user_impersonation`                                                         | Create a new delegated permission called user_impersonation.  Required value for this sample.      |

Reviewer: This is the on-behalf-of flow, so it's pretty common to see the `user_impersonation` string being used as the scope here, as it's not asking for any additional scopes, it's simply expecting the access token to be sufficient for the upstream call.  A lot of our examples do so it that way, as do a lot of Azure services as well, it's kinda a "thing" at this point, but honestly, I'm not a huge fan because that string almost looks like it's an "important" string (naming wise).  Conventions are a funny thing.

> | **API Permissions**               | `user_impersonation`                                                         | Create a new delegated permission called user_impersonation.  Required value for this sample.      |

Reviewer 2: OK, we'll roll with `user_impersonation` for now because this needs straightening out elsewhere.

The reason I asked this is because we have an open issue from a customer asking scope naming about convention:

> **Documentation says naming conventation is both access_as_user and Employees.Read.All**
> https://github.com/MicrosoftDocs/azure-docs/issues/84351

@jmprieur recommends `access_as_user` in https://github.com/MicrosoftDocs/azure-docs/issues/68540#issuecomment-756112758). It's also what we have in [Protected web API: App registration](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-protected-web-api-app-registration#exposing-delegated-permissions-scopes).

However, as you see in the quickstart, we use `resource.operation.constraint`.
 
None of them, however, specifies when to use which scope naming convention. I might sync up with @hpsin and see if we can get the story straight so I can update docs as needed.

> | **API Permissions**               | `user_impersonation`                                                         | Create a new delegated permission called user_impersonation.  Required value for this sample.      |

Reviewer 3: resource.operation.constraint is correct.  It just gets ignored because, as @ckittel points out, people have been ignoring it for a long time (including, of course, the defaults in Azure Portal). 

In Azure this is common because they use RBAC for authorization, and the operations, resources, and constraints are held there, in an expressive policy that supports delegation restrictions.  If Office were to use this we'd be in trouble.

> | **API Permissions**               | `user_impersonation`                                                         | Create a new delegated permission called user_impersonation.  Required value for this sample.      |

Reviewer 2: > If Office were to use **this** we'd be in trouble.

@hpsin Which *this* are you referring to here? :)

> | **API Permissions**               | `user_impersonation`                                                         | Create a new delegated permission called user_impersonation.  Required value for this sample.      |

Reviewer 3: If Office just used user_impersonation, or Graph did we'd be in trouble, because e.g. every user has permission to read their own email.  If Graph had just the one scope, you'd be unable to indicate "app Foo is allowed to read my email but not edit my files".  

Azure doesn't have the same conception of resources - you don't "have" files in Azure that are yours alone. They exist, and have their own ACLs, which extend to the user and to the app calling.

### 3. Modify your existing Web API

Update these settings to reference the new Web App

| App registration <br/> setting    | Value for this sample app                                                    | Notes                                                                                              |
|:---------------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|

> | **knownClientApplications**      | Client ID (UUID) of the application created in step 2.                       | Required value for this sample.                                                                    |

Reviewer 2: Use the exact UI element text (spaces, capiitalization, etc.).

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

### 4. Update code sample with app registration values

```javascript
// 'Directory (tenant) ID' in Azure portal

> const tenant = ''

Reviewer 2: Why is tenant a global? Should it not be in `config`?

Author: Updating

const config = {
  auth: {

    // 'Application (client) ID' of app registration in Azure portal - this value is a GUID
    clientId: '',

    // Client secret 'Value' (not the ID) from 'Client secrets' in app registration in Azure portal
    clientSecret: '',

    // Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
    authority: `https://login.microsoftonline.com/${tenant}`
  }
}
```

### 3. Install package(s)

To install Node.Js and MSAL libraries into your (virtual) environment:

```bash
npm install
```

## Run the application

```bash
node app.js
```

## Browse to the application

> Using postman or a similar app, navigate to **http://localhost:8080/me**

Reviewer: What web app?  It's just an API, so it would need to be Postman or curl or something similar that can pass custom HTTP headers.  What will the tutorial ask the user to do?

> Using postman or a similar app, navigate to **http://localhost:8080/me**

Reviewer 2: That question was added was before I looked at the code.

The title of the README says it's a **web app** and the app reg settings for a web app are specified. The app reg settings for the API say to use the registration for an **existing API**.

Thus, my confusion - everything indicated that this code sample at least _includes_ a web app because you're registering a web app.
